### ISSUE TITLE
Minor and patch release policy

### ISSUE THREAD

**Issue by Manishearth** [unicode-org/icu4x#2715]
Now that ICU4X 1.0 is out, we should determine what our policy is for minor and patch releases.

Some questions that should be answered:

 - How often should we make minor releases?
 - Can we make patch releases to individual components or must they bump every crate regardless of changes? What about minor releases?
 - What is the bar for a patch release?

cc @nordzilla @dminor @gregtatum @sffc @zbraniecki @robertbastian @hsivonen

**Comment by Manishearth** [comment]
Here's a strawman policy:

Minor releases occur on some regular cadence much like our pre-1.0 releases (strawman: once a quarter). Minor releases bump _all of ICU4X_ and contain changelogs, etc. They follow the whole ICU4X release process and are expected to have some level of coherent quality.

Patch releases can occur for individual components. The bar is higher than just landing a PR: to make a patch release, a separate PR must be made that creates the patch; and it must be approved by most of the team in a meeting or over Slack.

The patch release for, say, icu_calendar 1.0.1 may not occur at the same time as icu_timezone 1.0.1: the numbers are per-component.

Patch releases _may_ cover multiple components; for example perhaps icu_calendar and icu_datetime both need to be updated for a fix. Ideally they don't cover more than a couple; if a patch release needs to be global it is worth considering an out-of-cycle minor release instead.


MSRV can be bumped in minor but not patch releases.

**Comment by zbraniecki** [comment]
Input mental model:
1) `x.y.z` releases can be per component/util. No cadence, reactionary.
2) `x.y.0` releases should be coordinated, likely loosely periodic (train based)
3) `x.0` releases should be strictly coordinated, rare, more akin to Rust epochs. I'd love to start accumulating ideas for ICU4X 2.0 now but target it for at least a year ahead (and that's more frequent than I expect major releases to end up being, but maybe 1.0->2.0 is going to be more urgent due to influx of input).

**Comment by zbraniecki** [comment]
I also expect `x.y.0` to be more like current ICU4C/J releases, so 71->72->73 will be more like 1.1->1.2->1.3.
What I envision for `2.0` is unlike anything that ICU4C/J has as it allows for API removal.

**Comment by Manishearth** [comment]
Yeah, that seems to match my proposal above.

I didn't include guidance for major releases, but yes, I figure we should have a 2.0 in a year and after that major releases become more rare (though still not unexpected).

**Comment by sffc** [comment]
A question we should answer at the same time is how we deal with branches. Proposal: introduce a `1.x` branch; most work takes place on `main`, but individual commits can be cherry-picked to `1.x` to make patch releases. When we're ready for a full minor release, we merge all of `main` into `1.x`. API-breaking changes should not land even on `main`, except in experimental components, until we're getting ready for the `2.x` release stream.

I expect 2.0 in mid to late 2023, and from that point forward, once every 2-3 years.

> MSRV can be bumped in minor but not patch releases.

Seems like we'd want to bump MSRV in major releases. I do know that in Node.js at least, it's precarious to update all your dependencies because they might break on your Node.js version. I also think it's just generally a good practice to not use bleeding-edge features in stable software, because it takes time for people to update their toolchains, etc. Not everyone has the luxury of `rustup`.

**Comment by Manishearth** [comment]
@sffc 

So that I udnerstand your proposal, you are proposing the following, right?

We follow the proposal I had for patch releases, _except_ that patch releases happen off of a separate branch so that `main` stays clean, and we _only_ include the commits that triggered the patch in the patch release (instead of just "whatever has happened so far")

I think that's fine, except I'd recommend that for minor releases we merge `1.x` into `main` not vice versa: I think it's good for important releases to all be found on `main` without any branched history. In the ideal case, such a merge ought to not have any actual diff. And then we can merge `main` _back_ into `1.x`.

We'd also need to enable CI on 1.x, which seems easy enough.


> Seems like we'd want to bump MSRV in major releases

I'm fine with that as an MSRV policy. I would like to see if we can get away with using the workspace cargo versioning stuff (which AIUI will affect the compiler we use to build things but not the MSRV due to it getting normalized on publish, but it makes it harder to ensure MSRV)

**Comment by sffc** [comment]
Yes, and for minor releases we basically bring `1.x` and `main` back into sync with each other. Seems to make sense to merge `1.x` into `main` and then fast-forward `1.x` to be the same as `main`, and then the two branches can start diverging again.

**Comment by sffc** [comment]
If we don't want to deal with the hassle of maintaining multiple branches, then we could just do the "release everything that's landed so far" approach, except that it would be hard to enforce that there are no non-patch-level changes happening then.

I don't think we'd need a new CI job; people would make PRs directly against 1.x, and all the normal CI jobs should just work the same.

**Comment by hsivonen** [comment]
What's been written so far in this thread doesn't say anything about compatibility of data for a newer code release with past code releases or compatibility of data for a past code release with newer code releases.

I note that it's conceivable that there can be different levels of compat: Old code ignoring some aspect of newer data while producing results no worse than with the old data or new data being incompatible with old code in a way that would yield worse results with old code than old data.

(In general, I worry about promising too much about the benefit of data compat across versions and ending up with data design mistakes that theoretically cannot be fixed even if in practice affected parties could update both code and data.)

**Comment by hsivonen** [comment]
Also: Is it good to constrain all the crates to the same major version? That is, if one crate needs a breaking change, does it make sense to signal a semver break for the crates that don't need a semver break?

**Comment by Manishearth** [comment]
> What's been written so far in this thread doesn't say anything about compatibility of data for a newer code release with past code releases or compatibility of data for a past code release with newer code releases.


I think that's because this had been decided beforehand, mostly.

The general idea is that data keys are perma-stable and get versioned separately. We have been following the guidance of:

 - New data must work with old code 
   - old code _may_ ignore parts of new data if it needs to; e.g. the properties code is deliberately written so that new backends can be datagenned in the future. Our design usually requires this be a deliberate opt in for structs, though, IIRC)
 - Old data must work with new code

_Assuming_ the data key does not change.


If the data key does change, all bets are off, however we may mitigate impact by Introducing compat adapters that can be added to the data provider which can e.g. load old data on new code. This way people with strong control over their data pipeline can just use new code / new data, and people without strong control can use the adapter in case they get old data.

------

Pertinent to this discussion, I would say that patch versions should _never_ introduce new data versions, minor versions _can_ but should avoid it if possible.

**Comment by sffc** [comment]
Right; data stability has been discussed previously; see the following design doc:

https://docs.google.com/document/d/1yg_2l5FFo0aAuNi4jpgcIhIYjHqJyUoJWtMduyQ0vR8/edit

Basically:

1. Don't change the serialized form of currently existing data structs.
2. New data structs may be added, but any pre-existing `_with_any_provider` and `_with_buffer_provider` constructors must be able to continue reading data from the old data structs back to when the constructor was first introduced, within the same major version number.
3. To generate new data files supporting all ICU4X versions within a certain major version number, all data structs should be included, including obsolete ones. It is admissible to drop them upon upgrading to a new major version.

**Comment by Manishearth** [comment]
So right now we tag whole-ICU releases as `icu@.x.0`

We also are doing:

 - "major" bumps for pre-1.0 experimental deps
 - occasional out-of-cycle major, minor, and patch releases for utils
 - patch releases for ICU crates

We should tag these as much as possible, especially for cases like https://github.com/unicode-org/icu4x/pull/3091 where the commit will not be available in `main`.



Personally I'd like such "not the whole enchilada" releases to be under some "folder" in tags, perhaps `patches/foobar@x.y.z` or `individual/foobar@x.y.z`

We're not sure on the name to pick.

Also I would _like_ to make it easy to do the tagging when we bump utils (etc) so it's hard to miss.

**Comment by Manishearth** [comment]
Consensus between @sffc, @robertbastian, and @Manishearth: use `ind/<cratename>@x.y.z`

**Comment by sffc** [comment]
I believe that all of the questions in the OP are answered; so this just needs docs.
